rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -500,7 +515,8 @@ impl MessageStore<CanisterInput> for MessageStoreImpl {
         self.pool.get(reference).is_none()
             && (reference.class() != Class::BestEffort
                 || reference.kind() != Kind::Response
-                || !self.shed_responses.contains_key(&reference))
+                || (!self.expired_callbacks.contains_key(&reference)
+                    && !self.shed_responses.contains_key(&reference)))
```

Comment:
I switched it around to say `pool.get(_).is_none() && !(BestEffort && Response && (expired_callbacks.contains_key(_) || shed_responses.contains_key(_)))`. Hopefully that's clearer.

The alternative is to either write it like:
```rust
if self.pool.get(reference).is_some() {
    return false;
}

if reference.class == Class::BestEffort && reference.kind() == Kind::Response {
    self.expired_callbacks.contains_key(&reference)
        || self.shed_responses.contains_key(&reference)
} else {
    true
}
```
But that doesn't look more readable to me.

Or, we could do something like:
```rust
let is_in_pool = self.pool.get(reference).is_some();
let is_best_effort_response =
    reference.class == Class::BestEffort && reference.kind() == Kind::Response;
let is_expired_callback = self.expired_callbacks.contains_key(&reference);
let is_shed_response = self.shed_responses.contains_key(&reference);

return is_in_pool
    || is_best_effort_response && (is_expired_callback || is_shed_response);
```

But this would be both inefficient (every lookup is made every single time) and, again, not particularly readable either.

Reply:
I think the option you ended up with makes sense. Thank you.